File: api-java/src/main/java/com/example/generic/crud/GenericService.java
package com.example.generic.crud;

import com.example.response.GenericResponse;

import javax.inject.Inject;
import javax.ws.rs.NotFoundException;
import javax.ws.rs.core.Response;
import java.util.List;
import java.util.Optional;

public abstract class GenericService<T, ID, R extends GenericRepository> {

    @Inject
    R repository;

    public List<T> findAll() {
        return repository.findAll().list();
    }

    public T findById(ID id) {
        Optional<T> optional = repository.findByIdOptional(id);
        if (!optional.isPresent())
            throwNotFoundException();

        return optional.get();
    }

    public GenericResponse save(T entity) {
        repository.persist(entity);
        return GenericResponse.getGenericResponse("Registro(s) incluído(s) com sucesso.", Response.Status.CREATED.getStatusCode());
    }

    public GenericResponse update(T entity) {
        repository.getEntityManager().merge(entity);
        return GenericResponse.getGenericResponse("Registro(s) atualizado(s) com sucesso.", Response.Status.OK.getStatusCode());
    }

    public GenericResponse deleteById(ID id) {
        if (!repository.deleteById(id))
            throwNotFoundException();

        return GenericResponse.getGenericResponse("Registro deletado com sucesso.", Response.Status.OK.getStatusCode());
    }

    private void throwNotFoundException() {
        throw new NotFoundException("Registro não encontrado.");
    }

}

File: api-kotlin/src/main/kotlin/com/example/generic/crud/GenericService.kt
package com.example.generic.crud

import com.example.response.GenericResponse
import java.util.*
import javax.inject.Inject
import javax.ws.rs.NotFoundException
import javax.ws.rs.core.Response

abstract class GenericService<T, ID>(private val repository: GenericRepository<T, ID>) {

    fun findAll(): List<T> {
        return repository.findAll().list()
    }

    fun findById(id: ID): T {
        val optional: Optional<T> = repository.findByIdOptional(id)
        if (!optional.isPresent) throwNotFoundException()
        return optional.get()
    }

    fun save(entity: T): GenericResponse {
        repository.persist(entity)
        return GenericResponse("Registro(s) incluído(s) com sucesso.", Response.Status.CREATED.statusCode)
    }

    fun update(entity: T): GenericResponse {
        repository.entityManager.merge(entity)
        return GenericResponse("Registro(s) atualizado(s) com sucesso.", Response.Status.OK.statusCode)
    }

    fun deleteById(id: ID): GenericResponse {
        if (!repository.deleteById(id)) throwNotFoundException()
        return GenericResponse("Registro deletado com sucesso.", Response.Status.OK.statusCode)
    }

    private fun throwNotFoundException() {
        throw NotFoundException("Registro não encontrado.")
    }

}